cdc/generation.cc

```diff
@@ -119,176 +123,213 @@ const std::vector<token_range_description>& topology_description::entries() cons
- * shard and vnode pair there exists a stream whose token falls into this
- * vnode and is owned by this shard.
+/*
+ * Represents a vnode which is a (|start|, |end|] token range.
+ * It can be either a non wrap around range like (10, 100] or a wrap around
+ * range like (100, 10].
  *
- * It then builds a cdc::topology_description which maps tokens to these
- * found stream identifiers, such that if token T is owned by shard S in vnode V,
- * it gets mapped to the stream identifier generated for (S, V).
+ * The first example, (10, 100], has 90 tokens:
+ *     11, 12, ..., 99, 100
+ * while the second, (100, 10], has 2^64 - 90 elements
+ *     101, 102, ..., 2^63 - 1, -2^63, -2^63 + 1, ..., 9, 10
  */
-// Run in seastar::async context.
-topology_description generate_topology_description(
-        const db::config& cfg,
-        const std::unordered_set<dht::token>& bootstrap_tokens,
-        const locator::token_metadata& token_metadata,
-        const gms::gossiper& gossiper) {
-    if (bootstrap_tokens.empty()) {
-        throw std::runtime_error(
-                "cdc: bootstrap tokens is empty in generate_topology_description");
-    }
-
-    auto tokens = token_metadata.sorted_tokens();
-    tokens.insert(tokens.end(), bootstrap_tokens.begin(), bootstrap_tokens.end());
-    std::sort(tokens.begin(), tokens.end());
-    tokens.erase(std::unique(tokens.begin(), tokens.end()), tokens.end());
+struct vnode final {
+    dht::token start;
+    dht::token end;
+
+    /*
+     * Finds the first token in this vnode that belongs to shard |shard_idx|.
+     *
+     * If there is no token that belongs to shard |shard_idx| in this vnode,
+     * |end| is returned.
+     *
+     * The first token means the one that appears first on the ring when going
+     * from |start| to |end|.
+     * In non wrap around vnode like (10, 100], 'first token' simply means
+     * the smallest.
+     * In wrap around vnode like (100, 10], 'first token' is not always
+     * the smallest. For example, if only tokens 110 and 1 belong to
+     * shard |shard_idx| then token 110 is the first because it appears first
+     * when going from 100 to 10 on the ring.
+     */
+    dht::token find_first_token_for_shard(
+            const dht::sharder& sharder, size_t shard_idx) const;
+private:
+    /*
+     * Whether a vnode is a wrap around which means it contains both 2^63 - 1
+     * and -2^63.
+     */
+    bool is_wrap_around() const { return start >= end; }
+    /*
+     * Finds the first token in this vnode which is a wrap around.
+     *
+     * Such vnode has to be considered as two ranges.
+     * Range (|start|, 2^63-1] first and only when no token for shard
+     * |shard_idx| is found there, a second range of [-2^63, |end|] is searched.
+     */
+    dht::token find_first_token_for_shard_in_wrap_around(
+            const dht::sharder& sharder, size_t shard_idx) const;
+    /*
+     * Finds the first token in this vnode which is a wrap around.
```

Comment:
"which is not a wrap around" or "which is a non-wrap around"

Reply:
thanks.